File: Programmers/N-Queen.py
answer = 0
def solution(n):
    global answer
    visited = [False]*n
    board = [0]*n
    
    def check(x):
        for i in range(x):
            if abs(board[x]-board[i]) == x-i:
                return False
        return True
    
    def dfs(x):
        global answer
        if x == n:
            answer += 1
            return
        for y in range(n):
            if visited[y]:
                continue
            board[x] = y
            if check(x):
                visited[y] = True
                dfs(x+1)
                visited[y] = False
    
    dfs(0)
    
    return answer

File: Programmers/N-Queen.c
#include <stdio.h>
#include <stdbool.h>

int answer = 0;

bool check(int* board, int x) {
    for (int i = 0; i < x; i++) {
        if (abs(board[x] - board[i]) == x - i) {
            return false;
        }
    }
    return true;
}

void dfs(int* board, bool* visited, int x, int n) {
    if (x == n) {
        answer++;
        return;
    }
    for (int y = 0; y < n; y++) {
        if (visited[y]) {
            continue;
        }
        board[x] = y;
        if (check(board, x)) {
            visited[y] = true;
            dfs(board, visited, x + 1, n);
            visited[y] = false;
        }
    }
}

int solution(int n) {
    answer = 0;
    int board[n];
    bool visited[n];
    for (int i = 0; i < n; i++) {
        visited[i] = false;
    }
    dfs(board, visited, 0, n);
    return answer;
}
